cp1616/include/cp1616/cp1616_io_device.h

```diff
@@ -0,0 +1,380 @@
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int updateCyclicOutputData();
+
+  /**
+   * \brief Updates Input (from IO Controller perspective) data - write CP output data
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int updateCyclicInputData();
+
+  /**
+   * \brief This function is called after the PNIO_cbf_prm_end_ind callback has been
+   * called, it calls PNIO_initiate_data_write(), PNIO_initiate_data_read for the first
+   * time and PNIO_set_appl_state_ready
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int doAfterPrmEndIndCbf();
+
+  /**
+   * \brief This function is called after the PNIO_cbf_ar_indata_ind callback has been
+   * called, it calls PNIO_initiate_data_write() and PNIO_initiate_data_read()
+   *
+   * \return error_code (see pnioerrx.h for detailed description)
+   */
+          int doAfterIndataIndCbf();
+
+  /**
+   * \brief This function returns the index in the IO data array for the given submodule.
+   *
+   * \return index for given submodule
+   */
+          int GetSubmodNum(PNIO_UINT32 mod, PNIO_UINT32 sub);
+
+  /**
+   * \brief Data encapsulation for ar_info_ind_flag_
+   *
+   * \param value
+   */
+          void setArInfoIndFlag(int value);
+
+  /**
+   * \brief Data encapsulation for ar_info_ind_flag_
+   *
+   * \return ar_info_ind_flag_
+   */
+          int  getArInfoIndFlag();
+
+  /**
+   * \brief Data encapsulation for prm_end_ind_flag_
+   *
+   * \param value
+   */
+          void setPrmEndIndFlag(int value);
+
+  /**
+   * \brief Data encapsulation for ar_info_ind_flag_
+   *
+   * \return prm_end_ind_flag_
+   */
+         int  getPrmEndIndFlag();
+
+  /**
+   * \brief Data encapsulation for indata_ind_flag_
+   *
+   * \param value
+   */
+          void setIndataIndFlag(int value);
+
+  /**
+   * \brief Data encapsulation for indata_ind_flag_
+   *
+   * \return indata_ind_flag
+   */
+          int getIndataIndFlag();
+
+  /**
+   * \brief Data encapsulation for offline_ind_flag_
+   *
+   * \param value
+   */
+          void setOfflineIndFlag(int value);
+
+  /**
+   * \brief Data encapsulation for offline_ind_flag_
+   *
+   * \return offline_ind_flag_
+   */
+          int  getOfflineIndFlag();
+
+  /**
+   * \brief Data encapsulation for cp_session_key_
+   *
+   * \param value
+   */
+          void setCpSessionKey(PNIO_UINT16 value);
+
+  /**
+   * \brief Data encapsulation for cp_session_key_
+   *
+   * \return cp_session_key_
+   */
+          PNIO_UINT16 getCpSessionKey();
+
+  /**
+   * \brief Data encapsulation for cp_ar_number_
+   *
+   * \param value
+   */
+          void setCpArNumber(PNIO_UINT16 value);
+
+   /**
+    * \brief Data encapsulation for cp_ar_number_
+    *
+    * \return cp_ar_number_
+    */
+          PNIO_UINT16 getCpArNumber();
+
+
+  /**
+   * \brief CP handle obtained by init() function
+   */
+          PNIO_UINT32 cp_handle_;
```

Comment:
Perhaps you need to return `cp_handle_` by reference (i.e. `device_data_t&`).

The main problem I have with variables like `in_data_` being public is that their structure is complex (i.e. 3D array).  It requires the user to know which index is which.